### PR TITLE
kernel: Move LOCKED() macro to spinlock.h

### PR DESCRIPTION
**Pull request by finikorg**
Remove duplication in the code by moving macro LOCKED() to the correct
spinlock.h header.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

### CODE REVIEW

**Review by andyross — CHANGES_REQUESTED** [review]
So... I don't know that I'm comfortable with this yet. :)

I mean, yes, I wrote it and I love it.  But it's a tricky API: note that a break out of the loop or an early return will leave a dangling lock.  With GCC, there are extensions (the "cleanup" attribute -- essentially a way to write a C++-style destructor) to make this work reliably.  But I never investigated which toolchains support that or not.

Basically: I love using this down in the guts of the kernel, but I'm not sure I'm comfortable supporting this as an official API.

**Comment by ceolin** [comment]
> So... I don't know that I'm comfortable with this yet. :)
> 
> I mean, yes, I wrote it and I love it. But it's a tricky API: note that a break out of the loop or an early return will leave a dangling lock. With GCC, there are extensions (the "cleanup" attribute -- essentially a way to write a C++-style destructor) to make this work reliably. But I never investigated which toolchains support that or not.
> 
> Basically: I love using this down in the guts of the kernel, but I'm not sure I'm comfortable supporting this as an official API.

maybe just move it to `kernel_internal.h` then ? I liked the idea of having it in only one place, but I agree with you about making it public.

**Comment by finikorg** [comment]
> maybe just move it to `kernel_internal.h` then ? I liked the idea of having it in only one place, but I agree with you about making it public.

@andyross is this OK for you?

**Comment by ruuddw** [comment]
If there are concerns on using this, or additional guidelines are needed to e.g. ensure there is no break or early return performed, please add them clearly. Though it is sort of obvious to not exit a critical section with a spinlock kept - but with this abstraction it could be forgotten to release the lock in clean-up/exit code. Maybe include an example how to handle such a case: e.g. how to clean up in a 'goto err' construction or so?
kernel_internal.h seems the best place for this to me - how much user/application code would directly use spinlocks?

**Review by pabigot — CHANGES_REQUESTED** [review]
Looks good, just change the commit subject to reflect where it is now (not `spinlock.h`).

**Review by pabigot — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Works for me.  But if anyone wants to do the digging and figure out how to make this break/goto-safe in a portable way, I'd be super appreciative. :)

**Review by nashif — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: kernel/include/kernel_internal.h [hunk 1]
```diff
@@ -202,6 +202,11 @@ void z_mem_manage_init(void);
 /* Workaround for build-time page table mapping of the kernel */
 void z_kernel_map_fixup(void);
 
+#define LOCKED(lck) for (k_spinlock_key_t __i = {},			\
```
Comment (ruuddw) [inline kernel/include/kernel_internal.h:205]: A comment/warning on the safe use of the macro would be nice here: ensure there is no break/goto out of the body of the locked section to avoid a dangling spinlock.

**Review by ruuddw — APPROVED** [review]
ok, but still suggest to add a warning in code comments on how to use the macro.
